src/sentry/models/project.py

```diff
@@ -178,7 +181,15 @@ def save(self, *args, **kwargs):
                     reserved=RESERVED_PROJECT_SLUGS,
                     max_length=50,
                 )
-        super().save(*args, **kwargs)
+        try:
+            with transaction.atomic():
+                super().save(*args, **kwargs)
+        except IntegrityError:
+            self.snowflake_retry_counter += 1
+            if self.snowflake_retry_counter == 5:
+                raise Exception("Max allowed ID retry reached. Please try again in a second")
+            self.id = None
+            self.save(*args, **kwargs)
```

Comment:
I'd like to factor out the duplicated code between here and `Organization` into a mixin class. Lmk if you'd like details or to pair on it.